core/dbt/task/init.py

```diff
@@ -285,11 +296,25 @@ def run(self):
             if project_path.exists():
                 fire_event(ProjectNameAlreadyExists(name=project_name))
                 return
-            self.create_new_project(project_name)
-            profile_name = project_name
 
-        # Ask for adapter only if skip_profile_setup flag is not provided.
-        if not self.args.skip_profile_setup:
+            # If the user specified an existing profile to use, use it instead of generating a new one
+            user_profile_name = getattr(get_flags(), "PROFILE", None)
+            if user_profile_name:
+                # Verify it exists. Can't use the regular profile validation routine because it assumes
+                # the project file exists
+                raw_profiles = read_profile(profiles_dir)
+                if user_profile_name not in raw_profiles:
+                    print("Could not find profile named '{}'".format(user_profile_name))
+                    sys.exit(1)
+                profile_name = user_profile_name
+                profile_specified = True
+            else:
+                profile_name = project_name
+                profile_specified = False
+            self.create_new_project(project_name, profile_name)
+
+        # Ask for adapter only if skip_profile_setup flag is not provided and no profile to use was specified.
+        if not self.args.skip_profile_setup and not profile_specified:
```

Comment:
### User experience
From a user experience perspective, I don't think we need to raise an error if the specified profile is not found. I'd rather just create it anytime it doesn't exist.

### Implementation
I defer to whoever ends up being the code reviewer for this, but see below for some suggestions for refactoring.

The `run` method is long and has a lot of conditionals, which makes it harder to read. Refactoring this would make it easier to maintain in the future.

So I'd suggest refactoring this logic into its own method (similar to how `check_if_can_write_profile` is its own method). Maybe something similar this (completely untested!):

```python
def check_if_profile_exists(self, profile_name: Optional[str] = None) -> bool:
    profile_exists = False  # assume it doesn't exist unless proven otherwise
    user_profile_name = getattr(get_flags(), "PROFILE", None)
    profiles_dir = get_flags().PROFILES_DIR
    if user_profile_name:
        raw_profiles = read_profile(profiles_dir)
        profile_exists = user_profile_name in raw_profiles
    return profile_exists
```

Then this method can be applied in one or more places to use the specified profile if it exists (and create it otherwise).

Reply:
For user experience, I think more input is needed. The risk with creating the profile if it does not exist is the classic typo problem, where a misspelling creates a new profile instead of using the one the user actually wanted. The requirements said "existing profile" so I put in the check.

For implementation, the test is only done once currently, but shrinking a big method is usually a good idea. Refactored.

Reply:
Did you push the refactored code @ezraerb?

This is the most recent that I'm seeing:
<img width="284" alt="image" src="https://github.com/dbt-labs/dbt-core/assets/44704949/95f802a4-91a0-4550-827e-452d38d76ee8">

Reply:
Still working on a few other comments, so have not pushed yet.

Reply:
Changes have been pushed.

Reply:
My first instinct was to agree with @dbeatty10 on this point:

> From a user experience perspective, I don't think we need to raise an error if the specified profile is not found. I'd rather just create it anytime it doesn't exist.

I definitely appreciate the typo annoyance: `dbt init --profile defaut` would lead to the writing of a whole new `defaut` profile, when you _intended_ to use the existing `default` profile. In that case, it would be better to get an explicit error.

But it means that, when initializing a new project from scratch, you have exactly two options:
1. Do not pass `--profile` flag: Initialize a new project and a new profile, with the same name as your project.
2. Pass the `--profile` flag. It must match an existing profile.

If we took Doug's recommendation, there would be three options:
1. Do not pass `--profile` flag to initialize a new project and a new profile. The profile name will match your supplied project name (reasonable default behavior).
2. Pass the `--profile` flag, and its value matches an existing profile: Initialize a new project, do not write a new profile.
3. Pass the `--profile` flag, and its value doesn't match an existing profile: Initialize a new project and a new profile. The profile name will match what you passed into the `--profile` flag.

We could even take this one step further, and provide [the same flexibility that we offer when running `dbt init` within an existing project](https://github.com/dbt-labs/dbt-core/blob/b39eeb328c9a245f21b54fb48f2761ad03d4d6ad/core/dbt/task/init.py#L202-L205):
```
The profile <profile-name> already exists in /Users/jerco/.dbt/profiles.yml. Do you wish to overwrite it? [y/N]: N
```

---

I don't have strong feelings either way. As a heuristic to make this determination, I'm thinking about how there are two "modes" of `dbt init`:
1. **Interactive.** Likely first time using dbt. Need to set up everything.
2. **Programmatic.** Someone who has used dbt before, likely on this machine. Wants to skip the click interactivity and jump straight to `dbt init <project_name> --skip-profile-setup`, or `dbt init <project_name> --profile <existing_profile_name>`.

This `--profile` flag feels designed for persona / use case (2). The first user is less likely to want fine-grained control over exactly how the profile is being named — we should provide the easiest path from start to finish, with some sensible defaults along the way. And second user (slightly more experienced) is less likely to want the interactive walkthrough for setting up their profile.

Which is to say: I'm convinced enough that the behavior implemented in this PR is an acceptable user experience. We'll need to document the behavior in a new "Existing profile" section here: https://docs.getdbt.com/reference/commands/init